feat(sync): RN-1717: Sync bugs

chris-bes opened the pull request:
## Changes

- Fixing found bugs when testing for datatrak offline

## 🥞 Stack

1. #6289
2. #6446 👈

jaskfla commented on packages/database/src/server/changeHandlers/SyncLookupEntityRefresher.js:
```diff
@@ -0,0 +1,39 @@
+        SET updated_at_sync_tick = ?
+        WHERE id IN ${SqlQuery.record(entityIdsArray)}
+    `,
+      [1, ...entityIdsArray],
```
```suggestion
    await models.database.executeSql(
      `
        UPDATE entity
        SET updated_at_sync_tick = 1
        WHERE id IN ${SqlQuery.record(entityIdsArray)}
      `,
      entityIdsArray,
```

chris-bes commented on packages/database/src/core/migrations/20251005161529-addTombstoneTrigger-modifies-schema.js:
```diff
@@ -23,6 +23,9 @@ exports.up = function (db) {
       DECLARE
         current_tick bigint;
       BEGIN
+        IF ((SELECT value FROM local_system_fact WHERE key = 'syncTrigger') = 'disabled') THEN
```
Oh we have this for deterministic migrations in tamanu, which doesn't exist in tupaia yet. Maybe it's fine to remove it for now 👁️

jaskfla commented on packages/database/src/core/modelClasses/User/User.js:
```diff
@@ -304,14 +306,14 @@ export class UserModel extends DatabaseModel {
         return { ...obj, [key]: value };
       }, preferences);
       // If we change the selected project, we clear out the recent entities
       if (updatedPreferenceFields.project_id) {
-        updatedPreferenceFields.recentEntities = {};
+        updatedPreferenceFields.recent_entities = {};
       }
```
@chris-bes I’m not actually sure if this is still desirable or intended behaviour. This line of code is effectively dead because we only ever write to `recentEntities` (only in this line) and never read from it. There’s a chance that the comment is outdated and that this is no longer the right behaviour.

At some point, we switched from using `recentEntities` to `recent_entities`, and it changed from being an array to an object mapping country codes to entity IDs.
After that change, this line of code remained, but effectively `recentEntities` is deprecated.

Could be worth asking Andrew?

jaskfla replied:
This will be resolved by #6636

jaskfla commented on packages/datatrak-web/src/api/mutations/useEditUser.ts:
```diff
@@ -18,25 +28,32 @@ function camelToSnakeCase(camelCaseString: string): string {
+  return Object.fromEntries(
+    Object.entries(userDetails).map(([key, value]) => [camelToSnakeCase(key), value]),
+  );
+};
```
Actually a really good catch, but let’s descope from this project

jaskfla commented on packages/datatrak-web/src/database/DatatrakDatabase.ts:
```diff
@@ -12,6 +13,25 @@ export class DatatrakDatabase extends BaseDatabase {
+    // interprets input string in UTC. We want to treat these as floating times.
+    this.connection.client.pglite.parsers[types.TIMESTAMP] = (val: string) => val;
+  }
+
```
Pretty sure BugBot is conflating database class with model class. This doesn’t really make sense to me

jaskfla commented on packages/datatrak-web/src/database/DatatrakDatabase.ts:
```diff
@@ -12,6 +13,25 @@ export class DatatrakDatabase extends BaseDatabase {
+    // interprets input string in UTC. We want to treat these as floating times.
+    this.connection.client.pglite.parsers[types.TIMESTAMP] = (val: string) => val;
+  }
+
```
#6584

jaskfla commented on packages/datatrak-web/src/views/Sync/SyncPage.tsx:
```diff
@@ -46,18 +52,102 @@ const StyledLastSyncDate = styled(LastSyncDate)`
+    return () => {
+      syncManager.emitter.off('*', handler);
+    };
+  });
```
#6578 makes this moot

jaskfla commented on packages/database/src/core/modelClasses/SurveyResponse/upsertEntitiesAndOptions.js:
```diff
@@ -1,23 +1,7 @@
-const upsertEntities = async (models, entitiesUpserted, surveyId) => {
-  /** @type {import('@tupaia/database').SurveyRecord} */
-  const survey = await models.survey.findById(surveyId);
-  const dataGroup = await survey.dataGroup();
```
Intentional change

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -661,4 +751,36 @@ export class TaskModel extends DatabaseModel {
 
     return { dbConditions, dbOptions };
   }
+
```
`createAccessPolicyQueryClause` no longer returns null as of a0c673ab44c6c3bafe223b20feecac67ff6ae276

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -661,4 +751,36 @@ export class TaskModel extends DatabaseModel {
 
     return { dbConditions, dbOptions };
   }
+
```
Resolved in a0c673ab44c6c3bafe223b20feecac67ff6ae276

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -327,6 +339,10 @@ export class TaskModel extends DatabaseModel {
 
+    if (Object.keys(countryCodesByPermissionGroupId).length === 0) {
+      return null;
+    }
```
Resolved in a0c673ab44c6c3bafe223b20feecac67ff6ae276

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -661,4 +740,38 @@ export class TaskModel extends DatabaseModel {
 
     return { dbConditions, dbOptions };
   }
+
```
Resolved in a0c673ab44c6c3bafe223b20feecac67ff6ae276

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -661,4 +740,38 @@ export class TaskModel extends DatabaseModel {
 
     return { dbConditions, dbOptions };
   }
+
```
Resolved in a0c673ab44c6c3bafe223b20feecac67ff6ae276

jaskfla commented on packages/database/src/server/syncServerModelClasses/Answer.js:
```diff
@@ -0,0 +1,38 @@
+  async incomingSyncHook(records) {
+    const uploadPromises = records.map(async record => {
+      if (record.data.type === QuestionType.Photo) {
+        const newAnswerText = await s3Client.uploadImage(record.data.text, record.data.id);
```
da4f328760783d039eb51714aee47eef6e5be0f8

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -324,29 +336,34 @@ export class TaskModel extends DatabaseModel {
+          ${entries
+            .map(([, countryCodes]) => {
+              return `(survey.permission_group_id = ? AND entity.country_code IN ${SqlQuery.record(countryCodes)})`;
+            })
```
Technically correct, but this has never happened before. Descoping

jaskfla committed:
fix(database): `markSessionErrored` race condition

jaskfla committed:
fix(datatrak): RN-1490: blank country selector (#6608)

* unify label

* remove placeholder data

* define `useProjectCountryEntities`

* rerender `CountrySelector` on project change

* fix

* re-dedupe, revert prop rename, fix infinite render

* remove `useProjectCountryEntities`

* destructure

* fix bad usage

* remove redundant logic

jaskfla committed:
fix(database): `SyncSession. addInfo` race condition

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
feat(database): tiny type hint

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
type(centralServer): type hints

jaskfla committed:
fix(database): sync fact type mismatches (#6601)

* enum-ify sync facts

* fixes

* fix type mismatch

* type hints

* narrower types

* type hint

* fix incorrect type

* add missing string coercion

* tighten types

* declare SyncSnapshotAttributes.direction

* more accurate type

* make `incrementValue` return a number

* alphasort

* remove unused type

* safer usage

* deduplicate type

jaskfla committed:
refactor(sync): faster filter callback in `bulkUpdateForClient` (#6602)

* use set semantics

* iterate just once

* simplify

jaskfla committed:
tweak(datatrak): narrow type

jaskfla committed:
refactor(database): make `sync_session.info` JSONB

jaskfla committed:
tweak: various type hints

jaskfla committed:
tweak(centralServer): short circuit `EditUserAccounts`

jaskfla committed:
tweak(centralServer): remove unused import

jaskfla committed:
refactor: dedupilcate `snakeKeys` function

chris-bes committed:
add columnToMerge

chris-bes committed:
remove null

chris-bes committed:
clean up if stream throws error

jaskfla committed:
tweak(datatrakWeb): tighten `ClientSyncManager` types

jaskfla committed:
tweak: tighten types

jaskfla committed:
tweak(centralServer): remove unused import

jaskfla committed:
fix(sync): bad type

jaskfla committed:
tweak: type database `recordType` (and similar) symbols

jaskfla committed:
refactor(syncServer): don’t build unused map

jaskfla committed:
refactor(datatrakWeb): use safe bindings in `clearDatabase`

jaskfla committed:
tweak(database): more survey type hints

jaskfla committed:
fix(database): bad types

jaskfla committed:
refactor(database): use safe binding in `Survey` model

jaskfla committed:
tweak(database): type `FeedItem.createAccessPolicyQueryClause`

jaskfla committed:
style(datatrak): unify `FeedItem` and `Survey` model code style

jaskfla committed:
refactor(database): fetch only necessary columns

jaskfla committed:
tweak(database): organise exports

chris-bes committed:
missing await

chris-bes committed:
save more memory

chris-bes committed:
remove

jaskfla committed:
tweak(centralServer): declare `permissionsFilteredInternally` overrides as `const`s

jaskfla committed:
revert(database): restore missing exported types

This reverts commit b7377af808207678716c3f22f3f95c3d9f62bea9.

jaskfla committed:
test(database): work around Jest’s ES6 module limitation

chris-bes committed:
fix test

jaskfla committed:
test(database): fix same error with `SCHEMA`

jaskfla committed:
doc(sync): fix bad comment

chris-bes committed:
add some logs for debug

chris-bes committed:
add initial memory

chris-bes committed:
log error

chris-bes committed:
revert

chris-bes committed:
add more debug

jaskfla committed:
tweak(database): `deleteById` return type hint

jaskfla committed:
tweak(dabase): define model `join`s as `const`s

jaskfla committed:
tweak(database): narrower model types

jaskfla committed:
style(database): organise imports

jaskfla committed:
tweak(centralServer): throw nongeneric error

jaskfla committed:
doc(datatrakWeb): remove dead comment

chris-bes committed:
change batch size

chris-bes committed:
reduce batch size

chris-bes committed:
add more debug logs

chris-bes committed:
more logging

chris-bes committed:
fix batching

chris-bes committed:
more debug logs

chris-bes committed:
fixed debug

chris-bes committed:
more debug

chris-bes committed:
more debug

chris-bes committed:
update batch size

chris-bes committed:
more debug

chris-bes committed:
fix debug log

chris-bes committed:
updated batch size

jaskfla committed:
test: use valid Base64 URI in relevant unit tests

jaskfla committed:
test(centralServer): replace magic strings with enums

jaskfla committed:
fix(datatrak): RN-1752: `File` answer text format from offline uploads (#6623)

* neater typing

* shallower indenting

* const

* fix

* lint

* more readable

* remove dead comment

* lint

chris-bes committed:
fix batching

jaskfla committed:
tweak(datatrakWeb): narrower return type

jaskfla committed:
tweak(datatrakWeb): remove MediTrak reference from home view

jaskfla committed:
fix(datatrak): RN-1491: filter local surveys by both access policy and country (#6622)

* builds

* dedupe `ComparisonType` definition

* move existing logic to model layer

fr

* fix bad param

* format SQL

* fetch only necessary column

* fix

* typo

* types

* remove verbose comments

* types

* remove redundant error message definition

* fix use of incorrect models

* dedupe country ID clause

* remove dead import

* lazy typing

* safer

* use `orThrow`

* less blocking

* use `import type`

* more readable

jaskfla committed:
tweak: fix unsafe typing

chris-bes committed:
increase memory 2nd attempt

chris-bes committed:
increase

chris-bes committed:
increase memory

chris-bes committed:
vacuum

chris-bes committed:
create only

chris-bes committed:
no vacuum

chris-bes committed:
nginx + log

chris-bes committed:
more logs

chris-bes committed:
relaxedDurability

chris-bes committed:
upgrade

chris-bes committed:
test remove relaxedDurability

jaskfla committed:
tweak(centralServer): narrow `UserRecord` types

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

# Conflicts:
#	packages/database/src/core/DatabaseModel.js
#	packages/database/src/core/modelClasses/Project.js
#	packages/database/src/core/modelClasses/User/User.js
#	packages/datatrak-web-server/src/routes/UserRoute.ts

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
refactor(webConfigServer): more efficient `fetchHasPendingProjectAccess`

chris-bes committed:
revert unnecessary logs

chris-bes committed:
add logs for memory for incremental sync

chris-bes committed:
optional

jaskfla committed:
fix(database): restore missing type

jaskfla committed:
style(database): reflow awkward inline comment

chris-bes committed:
increase timeout

chris-bes committed:
increase timeout

chris-bes committed:
remove timeout

jaskfla committed:
fix(database): `debug_log` update race condition

chris-bes committed:
revert nginx

chris-bes committed:
add debug log

chris-bes committed:
add log

chris-bes committed:
remove users

jaskfla committed:
tweak(entityServer): narrower types

jaskfla committed:
fix(serverBoilerplate): unsafe property access

jaskfla committed:
refactor(entityServer): improve static type inference

jaskfla committed:
tweak(database): improve Task model & record type hints

jaskfla committed:
tweak(datatrak): clearer errors in bad mutation function calls

jaskfla committed:
tweak: custom error with unsafe accesses

jaskfla committed:
refactor(entityServer): fix unsafe access, short circuit

jaskfla committed:
tweak(tsmodels): bolster `Model` generic typing

jaskfla committed:
fix(datatrakWebServer): fetching a survey’s project

jaskfla committed:
tweak(datatrakWeb): fix query key collision

jaskfla committed:
tweak(serverUtils): use `node:` protocol import

jaskfla committed:
tweak(database): sync-related model instance method type hints

jaskfla committed:
tweak(constants): tighten `const` types

jaskfla committed:
refactor(datatrakWebServer): await promise

jaskfla committed:
style(datatrakWebServer): remove useless template

jaskfla committed:
refactor(datatrakWeb): more idiomatic query function

jaskfla committed:
fix(datatrak): RN-1490: make survey submission history available offline (#6625)

* tweak: fix unsafe typing

* fix bad type

* move to model layer

* use offline

* clean up diff

* alphasort

update schema

* more explicit

* better query key

* reuse object

* format

* group by primary key

* ensure

* fix

* again

jaskfla committed:
fix(datatrakWebServer): bad type

jaskfla committed:
tweak(datatrakWebServer): 404 when task not found

jaskfla committed:
tweak(datatrakWebServer): use const, interface

jaskfla committed:
refactor(datatrakWebServer): parallelise awaits

jaskfla committed:
fix(datatrakWeb): capitalisation of DataTrak

jaskfla committed:
tweak(datatrakWeb): refine update modal typography

jaskfla committed:
refactor: faster `ProjectModel.getAccessibleProjects` (marginally)

jaskfla committed:
tweak(tsmodels): tighten `MultiJoinItem` types

jaskfla committed:
tweak(centralServer): use `const`

jaskfla committed:
refactor(centralServer): short circuit if permission assertion fails

jaskfla committed:
refactor(centralServer): join in SQL

jaskfla committed:
refactor(datatrakWebServer): use `const`, `interface`

jaskfla committed:
tweak(datatrakWeb): improve query key

jaskfla committed:
refactor(datatrakWebServer): dedupe constant definition

jaskfla committed:
format(database): reflow SQL string

jaskfla committed:
tweak(database): more comprehensive SyncLookupQueryDetailsBuilder

jaskfla committed:
fix(tsmodels): bad type

jaskfla committed:
tweak(types): tighten `FeedItem` types

jaskfla committed:
format(centralServer): enforce code format

jaskfla committed:
style(centralServer): lint `importSurveyResponses`

chris-bes committed:
sync all entities

jaskfla committed:
refactor(dhisApi): use set semantics for cacheable endpoints

jaskfla committed:
refactor(centralServer): use `findOneOrThrow`, await promises

jaskfla committed:
tweak(centralServer): faster check for existing account

jaskfla committed:
tweak(centralServer): `emailVerification` type hints

jaskfla committed:
tweak(centralServer): `409 Conflict` when account exists

jaskfla committed:
refactor(centralServer): less I/O marking user as verified

jaskfla committed:
fix: hyphenation of ‘email’

jaskfla committed:
tweak(database): remove unused `@typedef`

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

jaskfla committed:
refactor(database): simplify `getAllProjectDetails` binding

jaskfla committed:
refactor(sync): faster save phase (#6604)

* use `partition`

* simplify detection of existing records

* fetch only ID

* one fewer iteration

* fix ambiguous projection

* more readable

* cleaner diff

* remove useless negation

* fix(database): `debug_log` update race condition

* use explicit cast

jaskfla committed:
Merge branch 'rn-1545-epic-datatrak-offline' into rn-1717-test-four

chris-bes committed:
Revert "sync all entities"

This reverts commit d074f8de078103195904893ab20bc1531b56101c.

chris-bes committed:
sync all entities

chris-bes committed:
clean up EntityHierarchyCacher

jaskfla committed:
refactor(centralServer): avoid needless fetch

jaskfla committed:
tweak(constants): more permission group name consts

jaskfla committed:
tweak(constants): dedupe object ID regex pattern

chris-bes committed:
repeatable read

jaskfla committed:
refactor(database): avoid needless `Date` instantiation

chris-bes committed:
remove repeatable read

chris-bes committed:
fix advisory lock

jaskfla committed:
refactor(centralServer): remove redundant predicate

jaskfla commented on packages/database/src/core/BaseDatabase.js:
```diff
@@ -190,15 +190,63 @@ export class BaseDatabase {
-    return await this.executeSql('SELECT pg_advisory_xact_lock(?)', [lockKeyInt]);
+    return await this.acquireAdvisoryLockByKey(lockKeyInt, {
+      ...options,
+      lockDescription: lockKey,
```
@chris-bes These two lines need to be flipped! Otherwise `lockDescription` will always be overwritten

jaskfla committed:
tweak(datatrakWeb): educate apostrophe

chris-bes committed:
Revert "fix advisory lock"

This reverts commit 962795ddf260e2ccd5ce998839495c2bb4f8e7ee.

chris-bes committed:
add autobind to tupaia database

jaskfla committed:
fix(database): bad `getTimezone` type and usages (#6637)

jaskfla committed:
refactor(datatrakWeb): set semantics for read-only question types

jaskfla committed:
refactor(datatrakWeb): extract survey route matchers, dedupe

jaskfla committed:
chore: dedupe type guards, delete unused

jaskfla committed:
refactor(centralSurvey): fetch only necessary column

jaskfla committed:
style(sync): remove useless templates

jaskfla committed:
tweak(apiClient): use const type

jaskfla committed:
format(datatrak): enforce code format

chris-bes committed:
fix(sync): RN-1717: Fix advisory lock causing db connection pool exhaustion (#6638)

* fix advisory lock

* comment

* address reviews

jaskfla committed:
refactor(datatrakWeb): faster check for unsynced data

jaskfla committed:
refactor(datatrakWeb): simplify project button label

jaskfla committed:
refactor(database): delegate ObjectID timestamp to bson-objectid

jaskfla committed:
tweak(database): `generateId` type hints

jaskfla committed:
fix(centralSever): RN-1711: slow email verification (#6634)

* tweak(centralServer): new accounts first when verifying email

* more readable

jaskfla committed:
fix(datatrakWeb): RN-1490: offline survey response history (#6632)

* move to model layer

* format(database): enforce code format

* tighten types

* bring offline

* dedupe

* remove dead import

* oops

* 🩹

* fetch only necessary data

* add missing sort

* fix

* more consistent

* add missing LIMIT clause

jaskfla committed:
chore(database): remove unused symbols

jaskfla committed:
style(database): no assignment in expression

jaskfla committed:
chore(database): dedupe instance method

jaskfla committed:
chore: remove suspicious `forEach` callback return

jaskfla committed:
refactor: avoid useless `Date` instance

jaskfla committed:
refactor(database): use optional chain

jaskfla committed:
refactor(database): use literal key

jaskfla committed:
refactor(database): use `indexOf`

jaskfla committed:
chore(utils): remove unused symbols

jaskfla committed:
refactor(utils): faster `calculateOuterBounds`

jaskfla committed:
chore: remove unused imports

jaskfla committed:
style: misleading variable name

chris-bes committed:
clean up properly after migration

jaskfla committed:
fix(database): undefined `handlerLock` in `TupaiaDatabase` subclass instances (#6641)

* fix

* comment

jaskfla committed:
refactor(datatrakWeb): define constants at top-level scope

jaskfla committed:
format(datatrakWeb): enforce code format

jaskfla committed:
format(datatrakWeb): reflow long comment

jaskfla committed:
tweak(datatrakWeb): use transaction to get user locally

jaskfla committed:
refactor(datatrakWeb):  use `Object.hasOwn`

jaskfla committed:
tweak(datatrakWeb): safer default local mutation function context type

jaskfla committed:
chore(datatrakWeb): use literal key

jaskfla committed:
chore(datatrakWeb): remove suspicious `let` declaration

jaskfla committed:
style(datatrakWeb): use optional chain

jaskfla committed:
chore(datatrakWeb): remove useless fragment

jaskfla committed:
refactor(datatrakWeb): remove spread from reducer

jaskfla committed:
refactor(datatrak): fix dubious activity feed Effect dependencies

jaskfla committed:
refactor(datatrakWeb): use only relevant primitive as dependency

jaskfla committed:
tweak(datatrakWeb): use transaction to save user locally

jaskfla committed:
tweak(datatrakWeb): remove bad selector

jaskfla committed:
tweak(datatrakWeb): clearer error message

jaskfla committed:
tweak(datatrakWeb): get entity descendants with transaction

jaskfla committed:
tweak(apiClient): use const type

jaskfla committed:
fmt(apiClient): enforce code format

jaskfla committed:
tweak(datatrakWeb): narrower inferred type

chris-bes committed:
fix issue sometimes there's no data showing after initial sync

chris-bes committed:
revert all debug logs

chris-bes committed:
revert debug logs 2

chris-bes committed:
remove logs

jaskfla committed:
tweak(apiClient): use const

chris-bes committed:
clean up datatrak auto update

jaskfla committed:
fmt(syncServer): format `createApp.ts`

jaskfla committed:
tweak(datatrakWeb): RN-1806: show loading spinner while logging out (#6643)

chris-bes committed:
clean up update

jaskfla committed:
refactor(datatrakWeb): use `find…OrThrow` helper methods

jaskfla committed:
tweak(uiComponents): define spinner once, balance message

jaskfla committed:
fix(centralServer): unsafe property access

jaskfla committed:
tweak: more readable lists in error messages

jaskfla committed:
tweak(centralServer): annotate abstract methods

jaskfla committed:
chore(centralServer): lint unused symbols

jaskfla committed:
refactor(centralServer): use optional chain

jaskfla committed:
test(serverUtils): update unit test to reflect expected behaviour

jaskfla committed:
fix(datatrak): public survey submission

jaskfla commented on packages/database/src/core/modelClasses/Task.js:
```diff
@@ -661,4 +745,38 @@ export class TaskModel extends DatabaseModel {
+              comparisonValue: null,
+            },
+          },
+        },
```
BugBot is misinterpreting

jaskfla committed:
tweak(datatrakWeb): clearer useLogin onSuccess error message

jaskfla reviewed (approved):
Stamping this one because the ultimate review will be on the base branch PR #6289

jaskfla committed:
tweak(datatrakWeb): memoise `SyncContext` value

jaskfla committed:
refactor(datatrak): decompose sync status listener hooks (#6585)

* type emitter

* extract into separate hooks

* convenience wrapper

* use new hook

* read `isSyncing` from `syncManager` every time

* refactor

* preserve existing behaviour

jaskfla committed:
fix(datatrakWeb): RN-1808: refreshing on sync page (#6651)

* type emitter

* extract into separate hooks

* convenience wrapper

* use new hook

* read `isSyncing` from `syncManager` every time

* refactor

* preserve existing behaviour

* fix

oops

* memoise event handler

* organise imports

* extract hook

jaskfla committed:
feat(datatrak): RN-1751: require up-to-date client to sync (#6644)

* enforce code format

* kinda sorta mostly there

* fixes

* update `attachApiClientToContext` interface

* use global axios `timeout`

fix bad export

* remove redundant custom handling

* refactors

* attach `X-Client-Version` to all Axios requests

* expore `X-Required-Client-Version` header

* display client version in GUI

* update comment

* remove debug logs

* refactor

* simplify

* delegate throw to backend

* simplify

* remove dead error handling

* remove redundant helper

* use semver

* un-propose `X-Required-Client-Version` header

* comment middleware side effect

* use CSS comment syntax

* trailing full stop

* test(syncServer): RN-1751: version compatibility test suite (#6647)

* try

* fix bad mock

* simplify

* delete stub test

* use `it.each`

* reorder

* refactor

* comment unrealistic cases

* fix

* typo

* proper reset before each

* mock next fn

fix

* less verbose

* dedupe type casts

* just mock it

* remove exclusive test

* fix test descriptions

* fix bogus `main` entrypoint

* update test mocks

* show app version text only in web app

jaskfla committed:
doc: list @tupaia/sync as a library package

chris-bes committed:
fix bug when mark record as deleted but record does not exist yet

chris-bes committed:
feat(datatrak): RN-1795: Notify user to reset data when permissions are changed (#6616)

* notify reset data

* use unknown

* clean up

* remove unused

* clean up

* clean up

* clean up

* add comment

* clean up

* clean up

* remove unused import

* add MIGRATIONS record type

* fix bad merge resolution

* type `countSyncSnapshotRecords` `parameters`

* fix build

* code format

* organise imports

* prefer interface

* lint Effect

* update wording for UX clarity

* `color-scheme` on banner

* code style

* fix hook rule violation

* code style

* smarter colouring

* extract

* add missing(?) mock

* fix browser login

* clearer errors

* refactor: avoid shadowing

* await in logout onSuccess

* use const

* types

* slightly faster logout

* move redirect into logout mutation fn

* fewer queries

* remove dead import

* fixes and tidy-ups

* refactor(datatrakWeb): simplify project button label

* faster permission change check

* this works

simplify

* improve type

* better wording

* Squashed commit of the following:

commit c7fe7fab59f3875d1151f2875d29fdbff378115e
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:23:11 2026 +1300

    type

commit 9668770c0ffb5ef2f74efb5fb2d94186e5c1467f
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 17:40:30 2026 +1300

    refactor

    simplify

    fix

    cherry pick

* Squashed commit of the following:

commit 37a04b8dd40570995d63d1bed8e83d2796d400ef
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 21:11:41 2026 +1300

    extract

    # Conflicts:
    #	packages/datatrak-web/src/components/ConfirmationModal.tsx
    #	packages/datatrak-web/src/hooks/useAbandonSurveyGuard.tsx

commit 50082806e4557102968e5d7d841134af77148172
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:35:07 2026 +1300

    fix

commit c7fe7fab59f3875d1151f2875d29fdbff378115e
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 18:23:11 2026 +1300

    type

commit 9668770c0ffb5ef2f74efb5fb2d94186e5c1467f
Author: Jasper Lai <33956381+jaskfla@users.noreply.github.com>
Date:   Mon Feb 16 17:40:30 2026 +1300

    refactor

    simplify

    fix

    cherry pick

* better prop names

* use logical props

* tighten types

* better naming

* fix

partial fix

redirect not working

* not quite

* use styled-components

* simplify and dedupe

remove dead import

* fix misleading symbol name

* clearer error message

* explicit refetch

* less cryptic error message

* tweak(datatrakWeb): use transaction to save user locally

* show only when online

* get permissionChanged with useDatabaseQuery

* clearer error message

* tweak(apiClient): use const

* fix bad resolution

* fix bad refetches after logout

* more readable

* feat(datatrak): RN-1608: Update notification UI change (#6646)

* update notification ui change

* fix color

* fix build

---------

Co-authored-by: Jasper Lai <33956381+jaskfla@users.noreply.github.com>

chris-bes committed:
fix reset data

chris-bes committed:
remove debug log in stream

chris-bes committed:
remove unused

chris-bes committed:
check if sync lookup is ready before allowing sync

chris-bes committed:
remove log

chris-bes committed:
export logMemory

chris-bes committed:
update yarn.lock

chris-bes committed:
fix test

chris-bes committed:
fix error

chris-bes committed:
fix test
